ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java

```diff
@@ -89,35 +90,72 @@ void shutdownGracefully() {
     }
   }
 
-  static class ReplyQueue implements Iterable<CompletableFuture<DataStreamReply>> {
+  static class ReplyQueue implements Iterable<DataStreamRequestEntry> {
     static final ReplyQueue EMPTY = new ReplyQueue();
 
-    private final Queue<CompletableFuture<DataStreamReply>> queue = new ConcurrentLinkedQueue<>();
+    private final Queue<DataStreamRequestEntry> queue = new ConcurrentLinkedQueue<>();
     private int emptyId;
 
     /** @return an empty ID if the queue is empty; otherwise, the queue is non-empty, return null. */
     synchronized Integer getEmptyId() {
       return queue.isEmpty()? emptyId: null;
     }
 
-    synchronized boolean offer(CompletableFuture<DataStreamReply> f) {
+    synchronized boolean offer(DataStreamRequestEntry f) {
       if (queue.offer(f)) {
         emptyId++;
         return true;
       }
       return false;
     }
 
-    CompletableFuture<DataStreamReply> poll() {
+    DataStreamRequestEntry poll() {
       return queue.poll();
     }
 
+    int size() {
+      return queue.size();
+    }
+
     @Override
-    public Iterator<CompletableFuture<DataStreamReply>> iterator() {
+    public Iterator<DataStreamRequestEntry> iterator() {
       return queue.iterator();
     }
   }
 
+  static class DataStreamRequestEntry {
+    private final CompletableFuture<DataStreamReply> replyFuture;
+    private final DataStreamRequest request;
```

Comment:
The request is for building the reply.  The reply should not be built in the client side.  Just complete the future exceptionally.

Also, the request has data which needs memory.

Since we don't need the request, we don't have to create the new DataStreamRequestEntry class.